### PR TITLE
LLVM 19 bugfixes + new CIs

### PR DESCRIPTION
**Pull request by franz**

### CODE REVIEW

File: lib/llvmopencl/linker.cpp [hunk 1]
```diff
@@ -458,7 +458,9 @@ static void replaceIntrinsics(llvm::Module *Program, const llvm::Module *Lib,
     }
   }
 
-  for (auto [Intrin, Repl] : EraseMap) {
+  for (auto It : EraseMap) {
```
Comment (pjaaskel) [inline lib/llvmopencl/linker.cpp:589]: Isn't C++17 OK? We support the oldest C++ standard supported by the oldest LLVM we support. I think LLVM moved to C++17 a while ago.

**Review by pjaaskel — APPROVED** [review]

File: lib/llvmopencl/linker.cpp (hunk 1 above)
Comment (franz) [reply lib/llvmopencl/linker.cpp:589]: LLVM 14&15 use C++14, the move only happened with LLVM 16.

File: lib/CL/clEnqueueSVMMemcpyRectPOCL.c [hunk 2]
```diff
@@ -81,7 +81,7 @@ pocl_svm_memcpy_rect_common (cl_command_buffer_khr command_buffer,
     const char *s = (const char *)src_ptr;
     char *d = (char *)dst_ptr;
     if (((s <= d) && (s + size > d)) || ((d <= s) && (d + size > s)))
-      POCL_RETURN_ERROR_ON (1, CL_MEM_COPY_OVERLAP, "overlapping copy \n");
+      POCL_RETURN_ERROR ( CL_MEM_COPY_OVERLAP, "overlapping copy \n");
```
Comment (linehill) [inline lib/CL/clEnqueueSVMMemcpyRectPOCL.c:84]: Nit:
```suggestion
      POCL_RETURN_ERROR (CL_MEM_COPY_OVERLAP, "overlapping copy \n");
```

File: lib/CL/clEnqueueWaitForEvents.c [hunk 3]
```diff
@@ -42,7 +42,7 @@ CL_API_SUFFIX__VERSION_1_0
   if (errcode != CL_SUCCESS)
     return errcode;
 
-  POCL_RETURN_ERROR_ON (1, CL_INVALID_OPERATION,
+  POCL_RETURN_ERROR ( CL_INVALID_OPERATION,
                         "clEnqueueWaitForEvents is not implemented\n");
```
Comment (linehill) [inline lib/CL/clEnqueueWaitForEvents.c:46]: Nit:
```suggestion
  POCL_RETURN_ERROR (CL_INVALID_OPERATION,
                     "clEnqueueWaitForEvents is not implemented\n");
```

File: lib/CL/clGetDeviceAndHostTimer.c [hunk 4]
```diff
@@ -39,7 +39,7 @@ CL_API_ENTRY cl_int CL_API_ENTRY POname (clGetDeviceAndHostTimer) (
   if (device->ops->get_synchronized_timestamps)
     return device->ops->get_synchronized_timestamps (device, device_timestamp,
                                                      host_timestamp);
-  POCL_RETURN_ERROR_ON(1, CL_INVALID_OPERATION, "Selected device "
+  POCL_RETURN_ERROR (CL_INVALID_OPERATION, "Selected device "
                        "does not support timestamp synchronization\n");
```
Comment (linehill) [inline lib/CL/clGetDeviceAndHostTimer.c:43]: Nit:
```suggestion
  POCL_RETURN_ERROR (CL_INVALID_OPERATION, "Selected device "
                     "does not support timestamp synchronization\n");
```

Comment (linehill) [reply lib/CL/clGetDeviceAndHostTimer.c:43]: More similar cases ahead.

File: lib/CL/devices/common.c [hunk 5]
```diff
@@ -1166,8 +1165,10 @@ pocl_check_kernel_dlhandle_cache (_cl_command_node *command,
 
   char module_fn[POCL_MAX_PATHNAME_LENGTH];
   int err = pocl_check_kernel_disk_cache (module_fn, command, specialize);
-  if (err)
+  if (err) {
+    POCL_UNLOCK (pocl_dlhandle_lock);
     return NULL;
+  }
```
Comment (linehill) [inline lib/CL/devices/common.c:1171]: Nit. Probably should be:
```suggestion
  if (err) 
    {
      POCL_UNLOCK (pocl_dlhandle_lock);
      return NULL;
    }
```

File: lib/CL/pocl_debug.h [hunk 6]
```diff
@@ -497,6 +497,16 @@ POCL_EXPORT
 
 #endif
 
+#define POCL_RETURN_ERROR(err_code, ...)                                    \
+    POCL_MSG_ERR2(#err_code, __VA_ARGS__); return err_code; \
+/*
+   do { \
+  } while (0)
+*/
```
Comment (linehill) [inline lib/CL/pocl_debug.h:505]: Dead code?

File: lib/CL/pocl_debug.h [hunk 7]
```diff
@@ -497,6 +497,16 @@ POCL_EXPORT
 
 #endif
 
+#define POCL_RETURN_ERROR(err_code, ...)                                    \
+    POCL_MSG_ERR2(#err_code, __VA_ARGS__); return err_code; \
```
Comment (linehill) [inline lib/CL/pocl_debug.h:501]:
```suggestion
#define POCL_RETURN_ERROR(err_code, ...)                                    \
    POCL_MSG_ERR2(#err_code, __VA_ARGS__); return err_code
```
I see there are `POCL_RETURN_ERROR(...);` calls. MSVC C compiler treats `;;`s as error.

File: lib/CL/pocl_llvm_wg.cc [hunk 8]
```diff
@@ -1704,7 +1704,12 @@ void populateModulePM(void *Passes, void *Module, unsigned OptL, unsigned SizeL,
       break;
     }
   }
-  ModulePassManager MPM = PB.buildPerModuleDefaultPipeline(Opt);
+  ModulePassManager MPM;
+  if (Opt == OptimizationLevel::O0) {
+    MPM = PB.buildO0DefaultPipeline(Opt);
+  } else {
+    MPM = PB.buildPerModuleDefaultPipeline(Opt);
+  }
```
Comment (linehill) [inline lib/CL/pocl_llvm_wg.cc:1712]: LLVM style nit:
```suggestion
  if (Opt == OptimizationLevel::O0)
    MPM = PB.buildO0DefaultPipeline(Opt);
  else
    MPM = PB.buildPerModuleDefaultPipeline(Opt);
```

File: lib/llvmopencl/Workgroup.cc [hunk 9]
```diff
@@ -267,6 +267,16 @@ bool WorkgroupImpl::runOnModule(Module &M, llvm::FunctionAnalysisManager &FAM) {
   // extra printf arguments.
   FunctionMapping PrintfCache;
 
+  // Remove the OptNone&NoInline keywords from all functions;
+  for (Module::iterator It = M.begin(), e = M.end(); It != e; ++It) {
+    Function &F = *It;
```
Comment (linehill) [inline lib/llvmopencl/Workgroup.cc:272]: I think, this simplifies to:
```suggestion
  for (Function &F : M) {
```

File: lib/llvmopencl/Workgroup.cc [hunk 10]
```diff
@@ -267,6 +267,16 @@ bool WorkgroupImpl::runOnModule(Module &M, llvm::FunctionAnalysisManager &FAM) {
...
+    if (!isKernelToProcess(F)) {
+      markFunctionAlwaysInline(&F);
+      F.removeFnAttr(Attribute::AlwaysInline);
+      continue;
```
Comment (linehill) [inline lib/llvmopencl/Workgroup.cc:276]: `continue` is redundant.

File: lib/llvmopencl/Workgroup.cc [hunk 11]
```diff
@@ -387,6 +397,30 @@ bool WorkgroupImpl::runOnModule(Module &M, llvm::FunctionAnalysisManager &FAM) {
...
+      FAM.clear(*F, "parallel.bc");
+      F->eraseFromParent();
+    }
+  }
+
+  for (auto Name : WorkgroupVariablesVector) {
+    GlobalVariable *GV = M.getGlobalVariable(Name);
+    if (GV && GV->getNumUses() == 0) {
+      GV->eraseFromParent();
+    }
```
Comment (linehill) [inline lib/llvmopencl/Workgroup.cc:415]: LLVM style nit:
```suggestion
    if (GV && GV->getNumUses() == 0)
      GV->eraseFromParent();
```

File: lib/llvmopencl/linker.cpp [hunk 12]
```diff
@@ -264,17 +322,92 @@ copy_func_callgraph(const llvm::StringRef func_name,
...
+        auto AllocatedType = AI->getAllocatedType();
+        const llvm::DataLayout &DL = Mod->getDataLayout();
+#if LLVM_MAJOR > 15
+        if (auto AllocaSize = AI->getAllocationSize(DL)) {
+          TotalSize += AllocaSize->getKnownMinValue();
+        }
+#else
+        if (auto AllocaSize = AI->getAllocationSizeInBits(DL)) {
+          TotalSize += AllocaSize->getKnownMinSize();
+        }
```
Comment (linehill) [inline lib/llvmopencl/linker.cpp:379]: LLVM style nit:
```suggestion
        if (auto AllocaSize = AI->getAllocationSize(DL))
          TotalSize += AllocaSize->getKnownMinValue();
#else
        if (auto AllocaSize = AI->getAllocationSizeInBits(DL))
          TotalSize += AllocaSize->getKnownMinSize();
```

File: lib/llvmopencl/linker.cpp [hunk 13]
```diff
@@ -264,17 +322,92 @@ copy_func_callgraph(const llvm::StringRef func_name,
...
+          return 0;
+        }
+
+        auto It = StackSizesMap.find(Callee);
+        if (It != StackSizesMap.end()) {
+          TotalSize += It->second;
+        } else {
+          TotalSize +=
+              estimateFunctionStackSize(Callee, Mod, CallChain, StackSizesMap);
+        }
```
Comment (linehill) [inline lib/llvmopencl/linker.cpp:401]: LLVM style nit:
```suggestion
        if (It != StackSizesMap.end())
          TotalSize += It->second;
        else
          TotalSize +=
              estimateFunctionStackSize(Callee, Mod, CallChain, StackSizesMap);
```

File: lib/llvmopencl/linker.cpp [hunk 14]
```diff
@@ -194,85 +246,157 @@ find_called_functions(llvm::Function *F,
...
+}
 
+/* Copy function F and all the functions in its call graph
+ * that are defined in 'from', into 'to', adding the mappings to
+ * 'vvm'.
+ */
+static int CopyFuncCallgraph(const llvm::StringRef FuncName,
+                               const llvm::Module *From, llvm::Module *To,
+                               ValueToValueMapTy &VVMap,
+                               PoclTypeRemapper *TypeMapper) {
```
Comment (linehill) [inline lib/llvmopencl/linker.cpp:300]: Style nit:
```suggestion
static int CopyFuncCallgraph(const llvm::StringRef FuncName,
                             const llvm::Module *From, llvm::Module *To,
                             ValueToValueMapTy &VVMap,
                             PoclTypeRemapper *TypeMapper) {
```

File: lib/llvmopencl/linker.cpp [hunk 15]
```diff
@@ -595,7 +718,11 @@ int link(llvm::Module *Program, const llvm::Module *Lib, std::string &Log,
 
   shared_copy(Program, Lib, Log, vvm);
 
-  removeDuplicateDbgInfo(Program);
+  if (StripAllDebugInfo) {
+    llvm::StripDebugInfo(*Program);
+  } else {
+    removeDuplicateDbgInfo(Program);
+  }
```
Comment (linehill) [inline lib/llvmopencl/linker.cpp:725]: LLVM style nit:
```suggestion
  if (StripAllDebugInfo)
    llvm::StripDebugInfo(*Program);
  else
    removeDuplicateDbgInfo(Program);
```

File: lib/CL/pocl_cl.h [hunk 16]
```diff
@@ -1174,6 +1184,11 @@ struct _cl_device_id {
      runtime/driver allocate the local space. */
   int device_alloca_locals;
 
+  /* Optional property. If the device uses stack and has stack size limitation
+   * (e.g. it does not convert all private memory to registers/spills),
+   * it should set up this property*/
+  size_t work_group_stack_size;
```
Comment (linehill) [inline lib/CL/pocl_cl.h:1190]: Perhaps should document what indicates the optionality (when this property is zero? When some macro is set?).

File: lib/llvmopencl/linker.cpp (hunk 15 above)
Comment (franz) [reply lib/llvmopencl/linker.cpp:725]: fixed in next PR

**Pull request merged**